File: spec/test_value_accessors.c
#include "testparser.h"
#include <assert.h>
#include <string.h>

int main()
{
    p_token_info_t token_info;
    char const * input = "42 f s";
    p_context_t * context;
    context = p_context_new((uint8_t const *)input, strlen(input));

    /* Default ptype value extracted with p_value_get(). */
    assert(p_lex(context, &token_info) == P_SUCCESS);
    assert(token_info.token == TOKEN_num);
    assert(p_value_get(&token_info.pvalue) == 42);

    /* Named ptype values extracted with p_value_get_XXX(). */
    assert(p_lex(context, &token_info) == P_SUCCESS);
    assert(token_info.token == TOKEN_flt);
    assert(p_value_get_float(&token_info.pvalue) == 1.5);

    assert(p_lex(context, &token_info) == P_SUCCESS);
    assert(token_info.token == TOKEN_str);
    assert(strcmp(p_value_get_string(&token_info.pvalue), "hello") == 0);

    p_context_delete(context);

    return 0;
}

File: spec/test_value_accessors.rs
use testparser::*;

fn main() {
    let mut c = p_context_new(b"42 f s");
    let mut ti = p_token_info_t::default();

    assert_eq!(P_SUCCESS, p_lex(&mut c, &mut ti));
    assert_eq!(TOKEN_num, ti.token);
    assert_eq!(42, p_value_get(&ti.pvalue));

    assert_eq!(P_SUCCESS, p_lex(&mut c, &mut ti));
    assert_eq!(TOKEN_flt, ti.token);
    assert_eq!(1.5, p_value_get_float(&ti.pvalue));

    assert_eq!(P_SUCCESS, p_lex(&mut c, &mut ti));
    assert_eq!(TOKEN_str, ti.token);
    assert_eq!("hello", p_value_get_string(&ti.pvalue));

    p_context_delete(c);
}
